Tweaks for MTC enhancements

landonreed opened the pull request:
This PR modifies and makes static/public a few internal methods for `Table` and `JdbcTableLoader` that are used in https://github.com/conveyal/datatools-server/pull/186.

evansiroky commented on src/main/java/com/conveyal/gtfs/loader/JdbcGtfsLoader.java:
```diff
@@ -259,27 +258,30 @@ private void registerFeed (File gtfsFile) {
      * It then creates a CSV reader for that table if it's found.
+     *
+     * This method is static so that it can be reused outside of JdbcGtfsLoader for getting CSV readers for zip files
+     * outside of the load procedure.
```
This comment makes me wonder if it may be more appropriate to have this method be within the Table class instead. Is that possible?

evansiroky commented on src/main/java/com/conveyal/gtfs/loader/ReferenceTracker.java:
```diff
@@ -0,0 +1,9 @@
+import java.util.HashSet;
+import java.util.Set;
+
+public class ReferenceTracker {
```
Please add a comment describing what the purpose of this class is.

evansiroky commented on src/main/java/com/conveyal/gtfs/loader/JdbcGtfsLoader.java:
```diff
@@ -485,6 +476,53 @@ private int loadInternal (Table table) throws Exception {
+     * @param fields array of fields (intended to be derived from the headers of a csv text file)
+     * @param fieldName string name of the field to find
+     */
+    public static int getFieldIndex(Field[] fields, String fieldName) {
```
Yeah, seeing how these are called from https://github.com/conveyal/datatools-server/pull/186/files#diff-b601a83ada6de294158d19894ceefe57, I think it might make more sense to put these methods in the Table class.

evansiroky replied:
Actually, this one doesn't have a Table argument, so maybe this should get put into the Field class.

evansiroky commented on src/main/java/com/conveyal/gtfs/loader/JdbcGtfsLoader.java:
```diff
@@ -485,6 +476,53 @@ private int loadInternal (Table table) throws Exception {
+     * @param table table from which to derive the key field
+     * @param fields array of fields (intended to be derived from the headers of a csv text file)
+     */
+    public static int getKeyFieldIndex(Table table, Field[] fields) {
```
Candidate for moving into Table class.

evansiroky commented on src/main/java/com/conveyal/gtfs/loader/JdbcGtfsLoader.java:
```diff
@@ -485,6 +476,53 @@ private int loadInternal (Table table) throws Exception {
+     * For an array of field headers, returns the matching set of {@link Field}s for a {@link Table}. If errorStorage is
+     * not null, errors related to unexpected or duplicate header names will be stored.
+     */
+    public static Field[] getFieldsFromFieldHeaders(String[] headers, Table table, SQLErrorStorage errorStorage) {
```
Candidate for moving into Table class.

evansiroky commented on src/main/java/com/conveyal/gtfs/loader/JdbcGtfsLoader.java:
```diff
@@ -562,24 +600,12 @@ private void setFieldToNull(boolean postgresText, String[] transformedStrings, i
      */
-    public String sanitize (String string) throws SQLException {
+    public static String sanitize (String string, SQLErrorStorage errorStorage) {
         String clean = string.replaceAll("[^\\p{Alnum}_]", "");
```
I don't quite understand this regex. Can you make a comment for it?

landonreed replied:
@abyrd, can you provide a description of this? (I can add your comments to code).

abyrd replied:
Sorry for the undocumented regex, obviously these things should always have a comment. This removes all characters that are not alphanumeric or an underscore. The reasoning being that all GTFS field names are composed of only alphanumeric characters and underscores, and limiting all supplied strings to those characters should completely preclude SQL injection.

evansiroky commented on src/main/java/com/conveyal/gtfs/loader/Table.java:
```diff
@@ -870,75 +893,86 @@ public Table getParentTable() {
         return parentTable;
     }
 
+    public Set<NewGTFSError> checkReferencesAndUniqueness(String keyValue, int lineNumber, Field field, String value, ReferenceTracker referenceTracker) {
```
Needs javadoc

landonreed commented:
@evansiroky, thanks for the suggestions. I've addressed the comment gaps and moved methods to Table/Field. I don't know exactly what the regex does, but I've tagged @abyrd who wrote that method originally. If you'd like we can wait for his comments before merge.

landonreed commented:
:tada: This PR is included in version 4.3.0 :tada:

The release is available on [GitHub release](https://github.com/conveyal/gtfs-lib/releases/tag/v4.3.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
